### PR TITLE
#85 EA-2198 implemented validation of isDefinedBy URL

### PR DESCRIPTION
**Pull request by rgraf**
Implemented search URL validation and HTTP GET request to get positive response

### CODE REVIEW

File: set-search-api-client/src/main/java/eu/europeana/set/search/service/SearchApiClient.java [hunk 1]
```diff
@@ -38,4 +38,14 @@ public SearchApiResponse searchItems(String uri, String apiKey, boolean descript
...
     		throws SearchApiClientException;
+    
+    /**
+     * This method is used to validate service URL
+     * @param uri
+     * @param apiKey
+     * @return status code
+     * @throws SearchApiClientException
+     */
+    int validateUrl(String uri, String apiKey) throws SearchApiClientException;
```
Comment (gsergiu) [inline set-search-api-client/src/main/java/eu/europeana/set/search/service/SearchApiClient.java:49]: not needed, use search items method

File: set-search-api-client/src/main/java/eu/europeana/set/search/service/impl/SearchApiClientImpl.java [hunk 2]
```diff
@@ -181,6 +179,27 @@ JSONObject searchItems(String uri) throws SearchApiClientException {
 	}
     }
 
+    /* (non-Javadoc)
+     * @see eu.europeana.set.search.service.SearchApiClient#validateUrl(java.lang.String, java.lang.String)
+     */
+    public int validateUrl(String uri, String apiKey) throws SearchApiClientException {
```
Comment (gsergiu) [inline set-search-api-client/src/main/java/eu/europeana/set/search/service/impl/SearchApiClientImpl.java:185]: not needed, update searchItems method to handle responses with 0 items included

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 3]
```diff
@@ -365,6 +371,59 @@ private void validateControlledValues(UserSet webUserSet) throws RequestBodyVali
...
     
+    /**
+     * The value of isDefinedBy is validated (e.g. search URL https://api.europeana.eu/record/search.json?) 
+     * to point to the Search API.
+     * We make a GET request upon creation to see if the request returns a 200 (meaning is valid).
+     * @param webUserSet
+     * @throws ParamValidationException 
+     * @throws RequestBodyValidationException
+     */
+    private void validateSearchUri(UserSet webUserSet)
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:382]: please rename to validateIsShownBy

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 4]
```diff
@@ -365,6 +371,59 @@ private void validateControlledValues(UserSet webUserSet) throws RequestBodyVali
...
+     * @param webUserSet
+     * @throws ParamValidationException 
+     * @throws RequestBodyValidationException
+     */
+    private void validateSearchUri(UserSet webUserSet) 
+	    throws ParamValidationException, RequestBodyValidationException {
+	
+	int STATUS_OK = 200;
+
+	if (webUserSet.getIsDefinedBy() != null) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:387]: please use webUserSet.isOpenSet() instead

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 5]
```diff
@@ -365,6 +371,59 @@ private void validateControlledValues(UserSet webUserSet) throws RequestBodyVali
...
+            String queryUrl = getBaseSearchUrl(webUserSet.getIsDefinedBy());
+            if (!queryUrl.equals(searchUrl)) {	    
+    	        throw new ParamValidationException(UserSetI18nConstants.USERSET_VALIDATION_INVALID_SEARCH_URL,
+		    UserSetI18nConstants.USERSET_VALIDATION_INVALID_SEARCH_URL,
+		    new String[] { WebUserSetModelFields.IS_DEFINED_BY, queryUrl});
+            }
+        	
+    	    String apiKey = getConfiguration().getSearchApiKey();
+    	    int getStatus = 0; 
+    	    try {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:398]: please update the validation according to the last comments in the ticket

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java [hunk 6]
```diff
@@ -365,6 +371,59 @@ private void validateControlledValues(UserSet webUserSet) throws RequestBodyVali
...
+     */
+    private void validateSearchUri(UserSet webUserSet) 
+	    throws ParamValidationException, RequestBodyValidationException {
+	
+	int STATUS_OK = 200;
+
+	if (webUserSet.getIsDefinedBy() != null) {
+            String searchUrl = getBaseSearchUrl(getConfiguration().getSearchApiUrl());
+            String queryUrl = getBaseSearchUrl(webUserSet.getIsDefinedBy());
+            if (!queryUrl.equals(searchUrl)) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:390]: please change the order of variables to avoid NPE

**Review by gsergiu — CHANGES_REQUESTED** [review]
please implement reqested changes

File: set-common/src/main/resources/config/set.properties.template [hunk 7]
```diff
@@ -51,3 +51,6 @@ authorization.api.name=usersets
 
 # User set API version
 set.api.version=
+
+# search URL
+europeana.search.url=
```
[the remark this answers is not in the document]
Comment (rgraf) [reply set-common/src/main/resources/config/set.properties.template]: removed

File: set-search-api-client/src/main/java/eu/europeana/set/search/connection/HttpConnection.java [hunk 8]
```diff
@@ -46,6 +46,25 @@ public String getURLContent(String url) throws IOException {
         }
     }
    
+    /**
+     * This method is used for service URL validation
+     * @param url
+     * @return status code
+     * @throws IOException
+     */
+    public int getURLStatus(String url) throws IOException {
```
[the remark this answers is not in the document]
Comment (rgraf) [reply set-search-api-client/src/main/java/eu/europeana/set/search/connection/HttpConnection.java]: replaced with getURLContent

File: set-search-api-client/src/main/java/eu/europeana/set/search/service/SearchApiClient.java (hunk 1 above)
Comment (rgraf) [reply set-search-api-client/src/main/java/eu/europeana/set/search/service/SearchApiClient.java]: replaced with searchItems

File: set-search-api-client/src/main/java/eu/europeana/set/search/service/impl/SearchApiClientImpl.java (hunk 2 above)
Comment (rgraf) [reply set-search-api-client/src/main/java/eu/europeana/set/search/service/impl/SearchApiClientImpl.java]: searchItems already handles it with initial value -1

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 3 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: renamed

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 4 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: replaced with isOpenSet

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 5 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java:393]: validation updated according to last comments

File: set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java (hunk 6 above)
Comment (rgraf) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java]: order changed

**Pull request merged**